File: cloudflare/alife-app/src/types/eventOperations.ts
import type { LocalizedText } from './eventComposition'

export type EventOccurrence = { id: string; eventId: string; startUtc: string; endUtc: string; localDate: string; status: string; isLegacyBackfill: boolean }
export type EventTeamMemberStatus = 'invited' | 'accepted' | 'declined' | 'ended'
export type EventTaskStatus = 'todo' | 'inProgress' | 'blocked' | 'done' | 'cancelled'
export type EventTeamMember = { id: string; eventId: string; memberId: string; displayName: string; status: EventTeamMemberStatus; joinedUtc?: string | null; declinedUtc?: string | null; endedUtc?: string | null }
export type EventRoleAssignment = { id: string; roleRequirementKey: string; memberId: string; status: EventTeamMemberStatus; acceptedUtc?: string | null; declinedUtc?: string | null; endedUtc?: string | null }
export type EventTaskApprovalStatus = 'notRequired' | 'notSubmitted' | 'pendingReview' | 'approved' | 'returned'
export type EventTask = { assignmentStatus?: 'unassigned' | 'invited' | 'accepted' | 'declined'; assignmentRespondedUtc?: string | null; preparation?: LocalizedText; preparationUpdatedUtc?: string | null; preparationPublicationCandidate?: boolean; stage?: string; eventOccurrenceId?: string | null; reviewerMemberId?: string | null; approvalStatus?: EventTaskApprovalStatus; approvalRound?: number; sourceType?: string | null; sourceId?: string | null; id: string; eventId: string; workflowStepId?: string | null; title: LocalizedText; description: LocalizedText; assignedMemberId?: string | null; status: EventTaskStatus; isRequired: boolean; requiresApproval: boolean; isRestricted: boolean; dueUtc?: string | null; completedUtc?: string | null; eTag: string; dependencies: Array<{ id: string; dependsOnEventTaskId: string; dependencyType: string }>; blockers: Array<{ id: string; reason: string; resolvedUtc?: string | null }> }
export type EventTeamWorkspace = { enabledModules?: Array<{ moduleCode: string; label: LocalizedText }>; members: EventTeamMember[]; roles: EventRoleAssignment[]; tasks: EventTask[]; roleRequirements: Array<{ moduleCode?: string; eligibility?: string[]; separationFrom?: string[]; requirementKey: string; roleCode: string; minimum: number; recommended: number; maximum?: number | null }>; readinessBlockers: LocalizedText[]; canManage: boolean; canConfigure?: boolean }

export type EventProgramItem = { id: string; sessionId: string; title: LocalizedText; description: LocalizedText; sortOrder: number; startOffsetMinutes: number; durationMinutes: number; ownerMemberId?: string | null }
export type EventSession = { id: string; occurrenceId: string; title: LocalizedText; startUtc: string; endUtc: string; placeJson: string; leadMemberId?: string | null; status: 'draft' | 'confirmed' | 'cancelled'; items: EventProgramItem[] }
export type EventProgramme = { eventId: string; occurrenceId: string; eTag: string; sessions: EventSession[]; canManage: boolean }

export type EventAvailabilityStatus = 'unknown' | 'available' | 'unavailable' | 'preferNot'
export type EventRosterAssignment = { id: string; serviceSlotId: string; memberId: string; status: 'invited' | 'confirmed' | 'declined' | 'ended'; replacesAssignmentId?: string | null }
export type EventServiceSlot = { id: string; occurrenceId: string; sessionId?: string | null; programItemId?: string | null; zoneId?: string | null; roleCode: string; roleLabel?: LocalizedText | null; startUtc: string; endUtc: string; requiredCount: number; eligibilityCode: string; confirmedCount: number; assignments: EventRosterAssignment[]; myAvailability?: EventAvailabilityStatus | null; moduleCode?: string; candidateMemberIds?: string[]; isRosterCandidate?: boolean; canAssign?: boolean }
export type EventRoster = { eventId: string; occurrenceId: string; eTag: string; slots: EventServiceSlot[]; readinessBlockers: LocalizedText[]; canManage: boolean; canConfigure?: boolean }

export type EventRosterGroup = { roleCode: string; moduleCode: string; memberIds: string[]; eTag: string }

export type EventRosterBatchChange = { occurrenceId: string; slotId: string; occurrenceETag: string; candidateGroupETag: string; memberId: string | null; replacesAssignmentId?: string | null }
export type EventRosterPage = { page: number; pageSize: number; total: number; timeZone: string; canManage: boolean; canConfigure: boolean; isRecurring: boolean; defaultsVersion: number | null; defaultsETag: string;
  occurrences: Array<{ id: string; startUtc: string; endUtc: string; roster: EventRoster }>; groups: EventRosterGroup[]; people: Array<{ id: string; displayName: string }> }

export type EventTaskDetail = { canRespond?: boolean; canPrepare?: boolean;
  participants?: Array<{ id: string; name: string }>; task: EventTask; canManage: boolean; canSubmit: boolean; canWithdraw: boolean; canReview: boolean; history: Array<{ id: string; round: number; action: string; actorMemberId: string; reviewerMemberId?: string; snapshotJson: string; reason: string; createdUtc: string }> }

File: backend/src/Alife.Domain/Entities/EventOperations.cs
using Alife.Domain.Enums;

namespace Alife.Domain.Entities;

public sealed class EventTask
{
    public Guid Id { get; set; }
    public Guid EventId { get; set; }
    public Guid? WorkflowStepId { get; set; }
    public string TitleEn { get; set; } = string.Empty;
    public string TitleZh { get; set; } = string.Empty;
    public string DescriptionEn { get; set; } = string.Empty;
    public string DescriptionZh { get; set; } = string.Empty;
    public Guid? AssignedMemberId { get; set; }
    public EventTaskStatus Status { get; set; } = EventTaskStatus.Todo;
    public bool IsRequired { get; set; }
    public bool RequiresApproval { get; set; }
    public bool IsRestricted { get; set; }
    public DateTime? DueUtc { get; set; }
    public DateTime? CompletedUtc { get; set; }
    public Guid ConcurrencyToken { get; set; } = Guid.NewGuid();
    public DateTime CreatedUtc { get; set; }
    public DateTime UpdatedUtc { get; set; }

    public GroupEvent Event { get; set; } = null!;
    public EventWorkflowStep? WorkflowStep { get; set; }
    public Member? AssignedMember { get; set; }
    public ICollection<EventTaskDependency> Dependencies { get; set; } = [];
    public ICollection<EventTaskDependency> Dependants { get; set; } = [];
    public ICollection<EventTaskBlocker> Blockers { get; set; } = [];
}

public sealed class EventTaskDependency
{
    public Guid Id { get; set; }
    public Guid EventTaskId { get; set; }
    public Guid DependsOnEventTaskId { get; set; }
    public string DependencyType { get; set; } = "finishToStart";
    public DateTime CreatedUtc { get; set; }
    public EventTask EventTask { get; set; } = null!;
    public EventTask DependsOnEventTask { get; set; } = null!;
}

public sealed class EventTaskBlocker
{
    public Guid Id { get; set; }
    public Guid EventTaskId { get; set; }
    public string Reason { get; set; } = string.Empty;
    public Guid CreatedByMemberId { get; set; }
    public DateTime CreatedUtc { get; set; }
    public Guid? ResolvedByMemberId { get; set; }
    public string? Resolution { get; set; }
    public DateTime? ResolvedUtc { get; set; }
    public EventTask EventTask { get; set; } = null!;
    public Member CreatedByMember { get; set; } = null!;
    public Member? ResolvedByMember { get; set; }
}

public sealed class EventRosterAvailability
{
    public Guid Id { get; set; }
    public Guid ServiceSlotId { get; set; }
    public Guid MemberId { get; set; }
    public EventAvailabilityStatus Status { get; set; } = EventAvailabilityStatus.Unknown;
    public DateTime UpdatedUtc { get; set; }
    public EventServiceSlot ServiceSlot { get; set; } = null!;
    public Member Member { get; set; } = null!;
}

public sealed class EventRosterAssignment
{
    public Guid Id { get; set; }
    public Guid ServiceSlotId { get; set; }
    public Guid MemberId { get; set; }
    public Guid AssignedByMemberId { get; set; }
    public EventRosterAssignmentStatus Status { get; set; } = EventRosterAssignmentStatus.Invited;
    public Guid? ReplacesAssignmentId { get; set; }
    public DateTime? ConfirmedUtc { get; set; }
    public DateTime? DeclinedUtc { get; set; }
    public DateTime? EndedUtc { get; set; }
    public DateTime CreatedUtc { get; set; }
    public DateTime UpdatedUtc { get; set; }
    public EventServiceSlot ServiceSlot { get; set; } = null!;
    public Member Member { get; set; } = null!;
    public Member AssignedByMember { get; set; } = null!;
    public EventRosterAssignment? ReplacesAssignment { get; set; }
}
